trace/tracestate/tracestate.go

```diff
@@ -0,0 +1,182 @@
+// forward slashes /.
+//
+// Value is an opaque string up to 256 characters printable ASCII RFC0020 characters (i.e., the
+// range 0x20 to 0x7E) except comma , and =.
+type TracestateEntry struct {
+	key   string
+	value string
+}
+
+// Key returns the key of TracestateEntry
+func (te *TracestateEntry) Key() string {
+	return te.key
+}
+
+// Value returns the value of TracestateEntry
+func (te *TracestateEntry) Value() string {
+	return te.value
+}
+
+// Get retrieves value for a given key from Tracestate ts.
+// If the key is not found then false is returned with the value "" is returned.
+// If the key is found then true is returned with its value.
+func (ts *Tracestate) Get(key string) (string, bool) {
+	if len(ts.tracestateEntries) == 0 {
+		return "", false
+	}
+	for _, entry := range ts.tracestateEntries {
+		if entry.key == key {
+			return entry.value, true
+		}
+	}
+	return "", false
+}
+
+// TraceEntries returns a slice of TracestateEntry.
+func (ts *Tracestate) TraceEntries() []*TracestateEntry {
+	return ts.tracestateEntries
+}
+
+func (ts *Tracestate) remove(key string) {
+	for index, entry := range ts.tracestateEntries {
+		if entry.key == key {
+			ts.tracestateEntries = append(ts.tracestateEntries[:index], ts.tracestateEntries[index+1:]...)
+			break
+		}
+	}
+}
+
+func (ts *Tracestate) add(entry *TracestateEntry) error {
+	ts.remove(entry.Key())
+	if len(ts.tracestateEntries) >= maxKeyValuePairs {
+		return fmt.Errorf("Set failed: reached maximum key/value pairs limit of %d", maxKeyValuePairs)
+	}
+	ts.tracestateEntries = append([]*TracestateEntry{entry}, ts.tracestateEntries...)
+	return nil
+}
+
+func isValid(key, value string) bool {
+	return keyValidationRegExp.MatchString(key) &&
+		valueValidationRegExp.MatchString(value)
+}
+
+// NewTracestateEntry creates a TracestateEntry object with given key and value.
+// It returns error if either key or value is invalid.
+func NewTracestateEntry(key, value string) (*TracestateEntry, error) {
+	if isValid(key, value) {
+		return &TracestateEntry{
+			key:   key,
+			value: value,
+		}, nil
+	}
+	return nil, errors.New("invalid parameters")
+}
+
+func containsDuplicateKey(entries []*TracestateEntry) (string, bool) {
+	keyMap := make(map[string]int)
+	for _, entry := range entries {
+		if _, ok := keyMap[entry.Key()]; ok {
+			return entry.Key(), true
+		}
+		keyMap[entry.Key()] = 1
+	}
+	return "", false
+}
+
+// NewFromEntryArray creates a Tracestate object from an array of key-value pair.
+// nil is returned with with an error if
+//  1. If the len of the entries > maxKeyValuePairs
+//  2. If the entries contain duplicate keys
+func NewFromEntryArray(entries []*TracestateEntry) (*Tracestate, error) {
+
+	if entries == nil {
+		return nil, errors.New("Invalid parameter")
+	}
+	if len(entries) == 0 || len(entries) > maxKeyValuePairs {
+		return nil, fmt.Errorf("Invalid number of tracestateEntry (%d)", len(entries))
```

Comment:
In Go, error strings are not capitalized: 

fmt.Errorf("number of entries (%d) is larger than the max (%d)", len(entries), maxKeyValuePairs)